File: editor-script-check-dependencies/scripts/check_dependencies.py
#!/usr/bin/env python3
"""
Editor script to create a component for the provided resource
"""

import os
import json
try:
    from urllib.request import urlopen
    from urllib.request import Request
    from urllib.request import HTTPError
except ImportError:
    from urllib2 import urlopen
    from urllib2 import Request
    from urllib2 import HTTPError

try:
    from ConfigParser import SafeConfigParser as ConfigParser
except ImportError:
    from configparser import ConfigParser

import re
from distutils.version import StrictVersion
import time


def MyStrictVersion(v):
    m = re.search('^(\D|)+(.*)', v)
    if m:
        v = m.group(2)
    v = v or "0.0"
    return StrictVersion(v)


def get_dependencies(config_file):
    config = ConfigParser()
    config.read(config_file)
    deps = []
    i = 0
    while True:
        if not config.has_option("project", f"dependencies#{i}"):
            break
        deps.append(config.get('project', f"dependencies#{i}"))
        i += 1
    return deps


def get_header():
    config_file = os.path.join(os.getcwd(),".editor-script-settings", "editor-script-check-dependencies")
    header = {"User-Agent": "editor-script-check-dependencies"}
    if os.path.exists(config_file):
        config = ConfigParser()
        config.read(config_file)
        if config.has_option("Authenticate", "TOKEN"):
            header['Authorization'] = "token " + config.get('Authenticate', 'TOKEN')
    return header


def have_releases(request):
    """Check if a repository is using releases
    GET /repos/:owner/:repo/releases"""
    response = urlopen(request)
    if response.getcode() not in [200, 304]:
        return False
    response_content = response.read()
    response_content.decode('utf-8')
    json_response = json.loads(response_content)
    if json_response:
        return True
    return False


def get_latest_version(request):
    """Get the latest version
    GET /repos/:owner/:repo/releases/latest"""
    response = urlopen(request)
    if response.getcode() not in [200, 304]:
        return

    response_content = response.read()
    response_content.decode('utf-8')
    json_response = json.loads(response_content)

    version_string = json_response["tag_name"]
    try:
        return MyStrictVersion(version_string)
    except ValueError:
        return MyStrictVersion("0.0")

def compare_versions(dependency, old, new, project, url):
    """Print our version information"""
    if new > old:
        print("   {}/archive/{}.zip".format(url, new))
    else:
        if old == MyStrictVersion("0.0"):
            print("   no release version found.".format(project))
        else:
            print("    using latest version.".format(project))
    print("   using master.zip is not recommended.".format(project))


def get_latest_tags(request):
    """If a project is not using releases presume the latest tag is the latest release
    GET /repos/:owner/:repo/tags"""
    response = urlopen(request)
    if response.getcode() not in [200, 304]:
        return

    response_content = response.read()
    response_content.decode('utf-8')
    json_response = json.loads(response_content)
    tags = []
    for p in json_response:
        tags.append(p["name"])
    try:
        tags.sort(key=MyStrictVersion)
        if tags:
            return MyStrictVersion(tags[-1])
        else:
            return MyStrictVersion("0.0")
    except ValueError:
        return MyStrictVersion("0.0")


def check_rate_limit(header, dependencies):
    """Github rate limits the API calls, this checks if we are and prints some info
    GET /rate_limit"""

    request_url = "https://api.github.com/rate_limit"
    request = Request(request_url, headers=header)
    response = urlopen(request)
    if response.getcode() not in [200, 304]:
        return
    response_content = response.read()
    response_content.decode('utf-8')
    json_response = json.loads(response_content)

    limit = json_response["resources"]["core"]["limit"]
    remaining = json_response["resources"]["core"]["remaining"]
    reset = json_response["resources"]["core"]["reset"]

    if len(dependencies) > remaining:
        if "token" not in header:
            print("You are rate limited to {} dependency checks, add a 'token' to increase this limit".format(limit))
        print("You are rate limited until", time.ctime(reset))
    if remaining == 0:
        return True


def main():
    game_project = os.path.join(os.getcwd(), "game.project")
    header = get_header()
    dependencies = get_dependencies(game_project)

    if check_rate_limit(header, dependencies):
        # No need to continue if we are rate-limited
        return

    for dependency in dependencies:
        url = re.match("(.*)\/archive", dependency)
        if not url:
            continue
        url = url.group(1)
        project = url.split("/")[-1]
        owner = url.split("/")[-2]

        print("Checking '{}'...".format(project))

        # Collect the version from the url
        try:
            current_version = MyStrictVersion(dependency.split("/")[-1].replace(".zip", ""))
        except ValueError:
            print("    current version does not follow Semantic Versioning, info could be unreliable.")
            current_version = MyStrictVersion("0.0")

        # First check of there are any releases
        releases = "https://api.github.com/repos/{}/{}/releases".format(owner, project)
        request = Request(releases, headers=header)
        try:
            if have_releases(request):
                latest = "https://api.github.com/repos/{}/{}/releases/latest".format(owner, project)
                request = Request(latest, headers=header)
                latest = get_latest_version(request)
            else:
                # Get the latest version from tags if there are no releases
                releases = "https://api.github.com/repos/{}/{}/tags".format(owner, project)
                request = Request(releases, headers=header)
                latest = get_latest_tags(request)

            compare_versions(dependency, current_version, latest, project, url)

            if "Authorization" not in header:
                time.sleep(2)  # We are only allowed to do one call every 2 min if we are not using authorization
        except HTTPError as e:
            print("    HTTPError {}.".format(e.code))

        print("\n")


if __name__ == '__main__':
    main()

File: editor-script-check-dependencies/scripts/check_dependencies.lua
-- THIS SCRIPT IS ONLY A WORK IN PROGRESS, I PAUSED THE PROGRESS BECAUSE HTTP ISN'T SUPPORTED IN EDITOR SCRIPTS
-- ONLY KEEPING THIS SCRIPT HERE IF I CAN FIND A WAY TO USE HTTP

local lip = require "editor-script-check-dependencies.scripts.LIP"
local semver = require "editor-script-check-dependencies.scripts.semver"

local M = {}

local ok = "✔"
local error = "✖"
local warning = "!"

local function get_libraries()
	local libraries = {}
	local i = 0
	local game_project = lip.load("game.project")
	
	while true do
		local dep_url = game_project.project["dependencies#" .. i]
		if dep_url == nil then
			break
		end
		table.insert(libraries, dep_url)
		i = i +1
	end
	return libraries
end


local function format_result(result_table)
	local entries = {}
	for _, dependency in pairs(result_table) do
		local entry = {
			name=dependency.owner .. "/" .. dependency.project .. " (" .. dependency.current .. ")",
			latest=dependency.latest_url,
		}
		local current = dependency.current ~= nil and semver(dependency.current:gsub("[v]+", "")) or nil
		local latest = dependency.latest ~= nil and semver(dependency.latest:gsub("[v]+", "")) or nil
		if current == nil then
			entry.prefix = warning
		else
			if latest == nil then
				entry.prefix = ok
				entry.latest = "No remote version found."
			elseif current >= latest then
				entry.prefix = ok
				entry.latest = nil
			else
				entry.prefix = error
			end
		end
		table.insert(entries, entry)
	end
	return entries
end

local function print_result(result_table)
	print()
	print("editor-script-check-depedendencies")
	print(" " .. ok      .. "  The dependency is up to date.")
	print(" " .. error   .. "  The dependency is outdated.")
	print(" " .. warning .. "  The dependency is not using a semver supported release.")
	print()
	print()
	local formatted = format_result(result_table)
	for _, entry in pairs(formatted) do
		print(entry.prefix .. " " .. entry.name)
		if entry.latest then
			print("    " .. entry.latest)
		end
		print()
	end
end

local function get_github_header()
	local header = {["User-Agent"] = "editor-script-check-dependencies"}
	local config_file = ".editor-script-settings/editor-script-check-dependencies"
	local file_attrib = editor.external_file_attributes(config_file)
	if file_attrib.exists then
		local config = lip.load(file_attrib.path)
		local token = config.Authenticate.github_token
		if token then
			header['Authorization'] = "token " .. token
		end
	end
	return header
end

local function github_standard(url, result_table)
	local owner, project, version = url:match("[%w]*://github%.com/([%w-]*)/([%w-]*)[%w/.]*/([%w%._]*).zip")
	if owner == nil or project == nil or version == nil then
		return false
	end

	local headers = get_github_header()
	local request_url = "https://api.github.com/repos/" ..owner .."/" .. project .. "/releases/latest"
	local response = http.request(request_url, {method="GET", as="json", headers=headers})
	
	local latest_url, tag_name
	if response.status == 200 then
		tag_name = response["body"]["tag_name"]
		if tag_name then
			latest_url = "https://github.com/" .. owner .. "/" .. project .. "/archive/" .. tag_name .. ".zip"
		else
			latest_url = "No remote versions found"
		end
	elseif response.status == 404 then
		request_url = "https://api.github.com/repos/" ..owner .."/" .. project .. "/tags"
		response = http.request(request_url, {method="GET", as="json", headers=headers})
		if response.status == 200 then
			if response.body[1] ~= nil then
				tag_name = response.body[1].name
				latest_url = "https://github.com/" .. owner .. "/" .. project .. "/archive/" .. tag_name .. ".zip"
			end
		end
	end

	table.insert(result_table, {
		owner=owner,
		project=project,
		current_url=url,
		latest_url=latest_url,
		current=version,
		latest=tag_name,
	})
end

function M.main()
	local libs = get_libraries()
	local result_table = {}

	for _, w in pairs(libs) do
		local response = github_standard(w, result_table)
		if response == false then
			print("Url pattern or host not supported: " .. w)
		end
	end
	print_result(result_table)
end

return M